Ptm stoich

pcruzparri opened the pull request:
### Framing
Master wrote a single `Modification Info List` column: count-based occupancy per site, pooled across
every spectra file, computed at the tail of `CalculateSequenceCoverage()`. This replaces it with
per-sample-group occupancy columns, adds an intensity-weighted variant, and moves the calculation into
mzLib's `BioPolymerGroup` so digestion-product groups can reuse it later.

**Where to start reading.** Occupancy is computed in mzLib by
`BioPolymerGroup.PopulateSampleGroupResults()`, from the spectra files and FlashLFQ intensities handed
to each group by `PostSearchAnalysisTask.QuantificationAnalysis()` and its glyco counterpart.
`ProteinGroup.GetTabSeparatedHeader()` and `ToString()` render the columns. The FlashLFQ area split
that weights intensity occupancy is `PostSearchAnalysisTask.DistributeQuantifiedIntensities()`.

### Output change (breaking)

`Modification Info List` is gone. In its place, per sample group:

| column | meaning |
|---|---|
| `SpectralCount_<label>` | PSMs from that sample group |
| `Intensity_<label>` | FlashLFQ intensity, when quantification ran |
| `CountOccupancy_<label>` | modified / total PSM count, per site |
| `IntensityOccupancy_<label>` | modified / total FlashLFQ area, per site |

The per-site cell format also changed, from `#aa71[Oxidation on S,info:occupancy=0.50(1/2)]` to
`pos71[Oxidation on S,info:fraction=0.50(1/2)]`. Denominators are now per sample group rather than
pooled over the whole run, so values are not comparable to previous releases even where a site appears
in both.

**The column set is dynamic.** Anything parsing this file by column index will break — resolve by
header name. `Number of PSMs` and everything after it shift by the number of sample groups.

### The parts worth stopping on

**Counts are per spectra file, never per channel.**
A TMT reporter channel and a SILAC channel are inferred from one spectrum, not observed separately, so
neither has a spectral count of its own. Multiplex runs get one count column per file. SILAC's invented
per-label files (`sample(K+8.014).mzML`, which exist so FlashLFQ can separate channels) get a **blank**
count cell rather than `0`, because zero would assert a measurement that was never made. Intensity
columns are unaffected — those come from a dictionary, not from a path join, and that asymmetry is
behind most of the edge cases here.

**Intensity occupancy is weighted by FlashLFQ peak areas.**
Each quantified peptidoform's area is split across the PSMs that identified it in that file, and each
PSM carries its share. A peptidoform identified but never quantified contributes nothing rather than
entering as a zero. The divisor is the filtered PSM set FlashLFQ was given, not every PSM in the run —
dividing by the unfiltered count credited part of each area to spectra that occupancy never sums over.
On an 18-file HeLa run that shortfall reached 57% of sites, with the old denominator a median of 86% of
the corrected value. It largely cancels in the ratio, since both arms lose signal together: 3.3% of
sites move by more than 0.05 occupancy and 1.3% by more than 0.10.

**Occupancy survives every path that skips quantification.**
Count-based columns are populated before the multiplex return, the label-free return, the
experimental-design failure return, and before anything that can throw; the `catch` re-populates them so
a mid-quantification failure yields one uniform count-only schema instead of a mix. Glyco gets the same
treatment, including its individual-file protein reports, which are written before quantification would
ever run.

**Occupancy is computed twice on a label-free run.**
Once before the early returns, then again after quantification propagates, because assigning the
quantification state invalidates the first result. That is the cost of having the columns survive every
early return and any throw. Not profiled.

**`SpectralMatch` equality is reference equality.**
Matching the inherited `GetHashCode`, and matching master, which declared neither `Equals` nor
`GetHashCode`. The `HashSet<ISpectralMatch>` in `ProteinGroup` therefore behaves exactly as before.

### Also in this PR

- `ProteinGroup` extends mzLib's `BioPolymerGroup`, delegating quantification, sample grouping and
  occupancy; alias properties keep existing consumers working.
- Scan metadata moves into an immutable `ScanMetadata` record on `Ms2ScanWithSpecificMass` and
  `SpectralMatch`, so heavyweight scan objects can be released after scoring. `Num Experimental Peaks`
  is re-read after XCorr pre-processing rewrites the spectrum, including for every chimeric precursor
  sharing it.
- `SpectralMatch` implements `ISpectralMatch`.
- `IntensitiesByFile` and `FilesForQuantification` became copy-on-read adapters over the base
  properties, so callers moved from mutating the getter to build-locally-then-assign, hoisted to one
  read per call on the hot path.
- The copy constructor now carries `PeptideFdrInfo` and `IsobaricMassTagReporterIonIntensities`, which
  were being dropped by every clone including SILAC's.
- `ThreadSafeObservableCollection` supports safe iteration.
- Smaller fixes: variant description access and null-safety, a discarded `possibleVariantPsms` ordering
  result, and a stale `ExperimentalDesign.tsv` leaking between tests via shared `TestData/`.

### What was run

Full suite green at **1783 passed / 0 failed / 1 skipped**, against stock mzLib **1.0.585**. No mzLib
change is required to build or test this.

Each occupancy behaviour has a test verified in both directions — reverted the production change,
confirmed the test fails, restored it, confirmed it passes:

- `PostSearchCountOnlyMultiFileWritesOneSchema` — two files, disjoint proteins, quantification off
- `GetMs2ScansRefreshesPeakCountForEveryChimericPrecursor` — chimeric low-CID precursors
- `GlycoCountBasedOccupancySurvivesQuantificationOff` — both glyco gaps, combined and per-file
- `TestTmtProteinGroupsHaveCountColumnsButNoIntensityColumns` — one count column per file, none per channel
- `TestSilacQuantification` — invented channels blank, real files keep their counts

The area-splitting change is not reachable by any test in the suite — no peptidoform in the test data
mixes accepted and rejected PSMs — so it was validated instead by re-running the full 18-file HeLa
search and diffing the occupancy columns against the pre-fix output. Both controls (`CountOccupancy` and
the FlashLFQ `Intensity` columns) came back bit-identical across all 5,244 sites, so the change is
isolated. Sites where every covering form is modified are exactly invariant, as the algebra requires.

### Deliberately out of scope

- **SILAC intensity occupancy does not work** — every site reports `0.0000(0/0)`. FlashLFQ's file list is
  rewritten to synthetic per-label names while PSM paths stay real, so the area lookup misses.
  Pre-existing, not introduced here. Fixing it requires revisiting the divisor's PSM population at the
  same time, since SILAC swaps that list for clones.
- **A site with no measured intensity prints `0.0000(0/0)` rather than an empty cell**, which reads as a
  measured zero. 6.5% of sites (339 of 5,244) on our dataset. The fix is a per-arm evidence filter in
  mzLib's `FormatOccupancy`; it is not in 1.0.585.
- **`IntensityOccupancy_` columns are emitted whenever FlashLFQ produced results at all**, rather than
  when a share was actually assigned, so a run that quantified nothing still advertises the columns.

### Follow-ups

Builds and tests against the released mzLib 1.0.585, so nothing has to land first.

Two follow-ups do need an mzLib release before they can land: peptide-group (digestion-product)
occupancy reporting, and MBR features feeding into occupancy. Both will prompt a small refactor of the
occupancy pipeline introduced here.

pcruzparri committed:
Initial implementation of occupancy by intensity. Requires ptm_stoich branch of mzlib. Code is untestested. Have seen some discrepancies with the quantity of quantified mods and their indexing.

pcruzparri committed:
Updating the ProteinGroup.ModInfo writing

pcruzparri committed:
Saving latest work. Still need to rebase onto master and address bug (uninitialized object). commit wip.

pcruzparri committed:
temp save

pcruzparri committed:
fixing protein sequence issue. Also added a condition to check if variants can be written.

pcruzparri committed:
.

pcruzparri committed:
fixing changes during rebasing to reflect master code again.

pcruzparri committed:
fixes to tests and occupancy to pass said tests. need to do a final test before pr.

pcruzparri committed:
Fixes to code for remaining tests. TestModificationInfoListProteinGroupsOutput() was not consistent when run individually vs when run with all other tests. The ordering of the output was changing. Fixed to not look for the first data line but rather for the one with the right PG.

pcruzparri committed:
update + master merging fixes

pcruzparri committed:
cleaned output

pcruzparri committed:
Cleaning up mod info outputs

pcruzparri committed:
final fix so mods with 0 intensity do not get written.

pcruzparri committed:
made filter by common variable and common fixed boolean. Need to link it to gui options. default will be to still write those mods out.

pcruzparri committed:
merge master into ptm_stoich

pcruzparri committed:
Merge branch 'master' into ptm_stoich

pcruzparri committed:
merge master into ptm_stoich

pcruzparri committed:
merging

pcruzparri committed:
temp: integrating ScanMetadata

pcruzparri committed:
Merge branch 'ptm_stoich' of https://github.com/pcruzparri/metamorpheus into ptm_stoich

pcruzparri committed:
git commit fixing mzlib/mm modification change causing integration issue.

pcruzparri committed:
time stamping updates to proteingroup implementing biopolymergroup.

pcruzparri committed:
Merge branch 'ptm_stoich' of https://github.com/pcruzparri/MetaMorpheus into ptm_stoich

pcruzparri committed:
working with vignette data and spectralmatchandbiopolymergrouprefactoring branch

pcruzparri committed:
cleaning up mod pipeline plus minor qol recommendations from copilot.

pcruzparri committed:
integrate mzlib stoich update

pcruzparri committed:
merging master

pcruzparri committed:
fixing tests to expect the new column outputs

pcruzparri committed:
Merge branch 'master' into ptm_stoich

pcruzparri committed:
Merge branch 'master' into ptm_stoich

pcruzparri committed:
Test fixes and consistent column generation enforcement

pcruzparri committed:
Merge branch 'master' into ptm_stoich

pcruzparri committed:
fix merge conflicts

pcruzparri committed:
Merge branch 'master' into ptm_stoich

pcruzparri committed:
cleaning

pcruzparri committed:
fixing changes to intended perviously behavior (early returns on quant if exp design is no bueno) and meeting test expectations.

pcruzparri committed:
more tests

pcruzparri committed:
Minor update to thread safe observable collection meant to facilitate iteration through collection without worrying about threading.

pcruzparri committed:
reduce diff lines

pcruzparri committed:
Merge branch 'master' into ptm_stoich

Resolve conflict in SpectralMatch.cs:
- Keep ScanMetadata record approach from ptm_stoich branch
- Add PrecursorScanDeconvolutionScore as separate property (not in ScanMetadata record)

pcruzparri committed:
Remove default throw in DeconHostViewModel switch

Reverts the defensive default case that throws ArgumentOutOfRangeException
for unhandled DeconvolutionType enum values (FromFile, Multiple). These
values were added in mzLib 1.0.579 and are not yet supported by the GUI.

Keeping the switch without a default allows unknown enum values to fall
through silently, matching master behavior and fixing test failures in
DeconHostViewModelTests and MetaDraw tests that transitively create a
DeconHostViewModel.

This is intentionally left as out-of-scope for this PR.

pcruzparri committed:
Fix SILAC copy constructor dropping IsobaricMassTagReporterIonIntensities and PeptideFdrInfo

The protected SpectralMatch copy constructor (used by PeptideSpectralMatch.Clone
for SILAC light/heavy conversion) was missing copies of:
- IsobaricMassTagReporterIonIntensities (TMT/iTRAQ/diLeu reporter ions)
- PeptideFdrInfo (peptide-level FDR stats)

Both are now copied from the source PSM to the clone.

Adds TestSilacClonePreservesQuantAndFdrData regression test.

pcruzparri committed:
Merge branch 'master' into ptm_stoich

pcruzparri committed:
accidentally removed used modules

pcruzparri committed:
Merge branch 'master' into ptm_stoich

trishorts reviewed (commented):
Automated code review (local pipeline). Critical/Major findings are posted as inline comments on the relevant lines.

trishorts commented on MetaMorpheus/TaskLayer/SearchTask/PostSearchAnalysisTask.cs:
```diff
@@ -1308,9 +1322,9 @@ private void WriteVariantResults()
-                        if (culture.CompareInfo.IndexOf(variant.VariantCallFormatDataString.Description, "missense_variant", CompareOptions.IgnoreCase) >= 0)
+                        if (culture.CompareInfo.IndexOf(variant.Description, "missense_variant", CompareOptions.IgnoreCase) >= 0)
                         {
                             if (variant.VariantCallFormatDataString.ReferenceAlleleString.Length == 1 && variant.VariantCallFormatDataString.AlternateAlleleString.Length == 1)
```
**Major** - Variant null-safety fix is incomplete — VariantCallFormatDataString still dereferenced

**Problem:**
This PR changed the variant classification guard/lookup from `variant.VariantCallFormatDataString.Description` to `variant.Description` and added a `variant.Description.IsNotNullOrEmpty()` null guard, which signals that `VariantCallFormatDataString` (or its `Description`) can be null. However, inside the `missense_variant` branch the code still dereferences `variant.VariantCallFormatDataString.ReferenceAlleleString.Length` and `variant.VariantCallFormatDataString.AlternateAlleleString.Length` with no null check. If a variant has a non-empty `Description` but a null `VariantCallFormatDataString`, this throws a `NullReferenceException`. `WriteVariantResults()` is not wrapped in a try/catch, so the exception propagates and aborts the post-search analysis.

**Impact:**
Any database containing applied sequence variations that carry a missense `Description` but lack `VariantCallFormatDataString` data will crash the entire search at the variant-writing step, losing all results. The new guard gives a false sense of safety because the very field it was meant to protect against is still dereferenced unconditionally a few lines later.

**Suggested fix:**
Either gate the whole branch on `variant.VariantCallFormatDataString != null`, or read the allele strings through the same null-safe accessor the `Description` change relies on, falling back to the MNV path (or skipping) when the VCF data is unavailable. Ensure `Description` and the allele-length sub-classification draw from a consistent, non-null source.

**Affected test scenarios:**
- WriteVariantResults - missense variant with non-empty Description but null VariantCallFormatDataString - completes without NRE
- WriteVariantResults - SNV vs MNV classification - still distinguished correctly when VCF allele data is present
- WriteVariantResults - variant with null/empty Description - skipped without throwing

pcruzparri replied:
Good catch. That variant.Description change was a temporary patch I'd added while working through some mzLib/MetaMorpheus integration — it isn't actually needed here, since this branch is on the same mzLib version as master. I reverted it back to the original VariantCallFormatDataString.Description code, so the incomplete guard (and the NRE risk it introduced) is gone.

trishorts reviewed (commented):
Automated code review - 4 finding(s) at severity >= Major, posted as inline comments.

trishorts reviewed (commented):
batch test

trishorts commented on MetaMorpheus/EngineLayer/ProteinParsimony/ProteinGroup.cs:
```diff
@@ -182,7 +183,15 @@ public void GetIdentifiedPeptidesOutput(List<SilacLabel> labels)
+        public new string GetTabSeparatedHeader()
         {
             var sb = new StringBuilder();
             sb.Append("Protein Accession" + '\t');
```
**Major** - Per-row quant columns derived from each group's own SampleGroupResults can desync from the single header

**Problem:**
`GetTabSeparatedHeader()` emits the dynamic quant/occupancy columns by iterating `SampleGroupResults` and, per group, checking `group.HasIntensityData`. But the header is produced from one representative `ProteinGroup`, whereas `ToString()` re-derives the same columns from *each* `ProteinGroup`'s own `SampleGroupResults` and per-group `HasIntensityData`. The old base code derived columns from the globally-shared `FilesForQuantification.GroupBy(p => p.Condition)`, which was identical for every protein group and therefore could not desync. The new consumption pattern assumes `SampleGroupResults` count, ordering, and each group's `HasIntensityData` flag are uniform across all protein groups, but nothing in this file enforces that.

**Impact:**
If any protein group's `SampleGroupResults` differs in count/ordering, or a group has `HasIntensityData == false` while the header group had it `true` (e.g., a protein with no measured intensity in one sample), that row emits fewer/extra tab-separated cells than the header, silently shifting every downstream column and corrupting the TSV for parsers and MetaDraw.

**Suggested fix:**
Drive both header and rows from a single shared schema (e.g., the global sample-group list and a per-sample-group `HasIntensityData` decision computed once), rather than letting each `ProteinGroup` independently decide its own column set. Ensure the column shape is invariant across all rows and the header.

**Affected test scenarios:**
- ProteinGroupTest - two protein groups where one sample group has intensity for PG-A but not PG-B - header and both rows have identical column counts
- QuantificationTest - mixed-observation experiment - every protein row column count equals header column count

pcruzparri replied:
You're right that the header (built from the first group) and the per-group rows could drift apart in principle. In the normal pipeline every group gets IntensitiesByFile assigned (with zeros for unmeasured proteins), so the column schema is uniform in practice today. I agree it's still fragile. For now, I've added a guard test that builds multiple protein groups and asserts every row's column count matches the header. The actual header refactor (driving header + rows from one shared schema) I'll do in a separate PR since it'd be best to address that at the BioPolymerGroup (mzlib) level.

trishorts commented on MetaMorpheus/EngineLayer/ProteinScoringAndFdr/ProteinScoringAndFdrEngine.cs:
```diff
@@ -165,8 +165,8 @@ private List<ProteinGroup> DoProteinFdr(List<ProteinGroup> proteinGroups)
+                pg.BestPeptidePEP = pg.AllPsmsBelowOnePercentFDR.OfType<SpectralMatch>().Min(psm => psm.FdrInfo.PEP);
             }
 
             // pick the best for each paired accession based on filter type
```
**Major** - `.OfType<SpectralMatch>()` silently drops PSMs from best-QValue/PEP computation

**Problem:**
The only functional change in this file adds `.OfType<SpectralMatch>()` before `.Min(...)` on the `BestPeptideQValue` and `BestPeptidePEP` assignments, while `BestPeptideScore` on the line above still aggregates over the full `AllPsmsBelowOnePercentFDR` set. Because `AllPsmsBelowOnePercentFDR` is now exposed through the `BioPolymerGroup`/`ISpectralMatch` refactor, the cast was presumably added to reach `FdrInfo` (not on `ISpectralMatch`). But `OfType` *filters* rather than *casts*: any element that is not exactly a `SpectralMatch` is silently excluded from the QValue/PEP minimum, yet still included in `BestPeptideScore.Max`. The three "best peptide" metrics are then computed over inconsistent subsets of PSMs.

**Impact:**
If the set ever contains an `ISpectralMatch` that is not a `SpectralMatch`, `BestPeptideQValue`/`BestPeptidePEP` are computed from a different (smaller) population than `BestPeptideScore`, producing inconsistent sort keys in `SortProteinGroupsByFilterType` and incorrect target/decoy "picked" selection and protein-level FDR. If the filter removes *all* elements from a non-empty set, `.Min()` throws `InvalidOperationException` (sequence contains no elements), aborting the whole scoring engine.

**Suggested fix:**
Use `.Cast<SpectralMatch>()` (which throws loudly on an unexpected type, surfacing the real bug) instead of `.OfType<SpectralMatch>()`, or downcast a single shared filtered list once and reuse it for all three metrics so Score, QValue, and PEP are computed over the same PSM population. If non-`SpectralMatch` elements are genuinely possible, handle that case explicitly rather than letting it skew aggregates silently.

**Affected test scenarios:**
- DoProteinFdr - protein group whose PSM set mixes `SpectralMatch` with another `ISpectralMatch` impl - QValue/PEP/Score computed over identical population
- DoProteinFdr - protein group with non-empty PSM set but no `SpectralMatch` after filter - does not throw; metrics well-defined
- ScoreProteinGroups/DoProteinFdr - all-`SpectralMatch` set (normal case) - best metrics unchanged from pre-refactor behavior

pcruzparri replied:
Agreed. Switched to a single Cast<SpectralMatch>().ToList() that's reused for Score, QValueNotch, and PEP, so all three are computed over the same PSM set and an unexpected type throws loudly instead of being silently dropped. Parsimony/FDR tests still pass.

trishorts commented on MetaMorpheus/GuiFunctions/Util/ThreadSafeObservableCollection.cs:
```diff
@@ -0,0 +1,78 @@
+
+    protected override void OnCollectionChanged(NotifyCollectionChangedEventArgs e)
+    {
+        Dispatcher dispatcher = Application.Current?.Dispatcher;
```
**Critical** - Synchronous dispatcher.Invoke raised while holding _lock can deadlock

**Problem:**
Every mutation (`InsertItem`, `RemoveItem`, `SetItem`, `MoveItem`, `ClearItems`) acquires `_lock` and then calls the `base.*` method, which synchronously raises `OnCollectionChanged`. The overridden `OnCollectionChanged` calls `dispatcher.Invoke(() => base.OnCollectionChanged(e))`, a *blocking* marshal to the UI thread — and it runs while `_lock` is still held. If the UI thread is concurrently blocked trying to acquire `_lock` (e.g., enumerating via the new `GetEnumerator`, or performing its own mutation), the background thread cannot complete its `Invoke` because the UI thread is parked on the lock, and the UI thread cannot release because it is waiting for the lock the background thread holds — a classic two-party deadlock.

**Impact:**
The application hangs (frozen UI) under exactly the concurrent read/write scenario this class was created to make safe. Because the class is intended for background-thread mutation with UI-thread binding, this is a high-probability hang in production.

**Suggested fix:**
Do not hold `_lock` across the change-notification raise. Capture/perform the mutation under the lock, then raise `OnCollectionChanged` outside the lock; or marshal notifications asynchronously (`BeginInvoke`) rather than synchronously, accepting that the dispatched delegate must not assume the collection is unchanged. Re-evaluate whether the lock must wrap the entire `base.InsertItem` call given that call also raises events.

**Affected test scenarios:**
- ThreadSafeObservableCollection - background thread Adds while UI thread enumerates under an active Dispatcher - completes without deadlock
- ThreadSafeObservableCollection - background thread Add while UI thread Insert - both complete, no hang

pcruzparri replied:
This whole ThreadSafeObservableCollection was out of scope for this PR — it was an attempt to fix a test flake that only showed up in my local environment, not in CI. I've reverted it out of the branch entirely, so the deadlock risk no longer exists here. If we decide we want a thread-safe collection later, I'll do it in its own PR and handle the locking/marshalling properly.

trishorts commented on MetaMorpheus/GuiFunctions/Util/ThreadSafeObservableCollection.cs:
```diff
@@ -0,0 +1,78 @@
+
+    // ── UI-thread marshalling ─────────────────────────────────────────────────
+
+    protected override void OnCollectionChanged(NotifyCollectionChangedEventArgs e)
```
**Major** - PropertyChanged notifications not marshalled to UI dispatcher

**Problem:**
The class overrides only `OnCollectionChanged` to marshal to the dispatcher. `ObservableCollection<T>` also raises `PropertyChanged` for `"Count"` and `"Item[]"` via `OnPropertyChanged`, which is left unmarshalled. When a mutation occurs on a background thread, the `PropertyChanged` event fires on that background thread directly. The class summary claims "CollectionChanged notifications are automatically marshalled to the UI dispatcher," but the equally cross-thread-sensitive `PropertyChanged` path is not handled.

**Impact:**
WPF bindings that listen to `Count` (or any property-change subscriber expecting UI-thread delivery) can receive notifications on a background thread, raising cross-thread `InvalidOperationException` or causing undefined binding behavior — defeating the thread-safety guarantee for those consumers.

**Suggested fix:**
Override `OnPropertyChanged` with the same dispatcher-marshalling logic used in `OnCollectionChanged`, or centralize the marshalling so both notification paths are covered consistently.

**Affected test scenarios:**
- ThreadSafeObservableCollection - background thread Add with a Count-bound subscriber under an active Dispatcher - PropertyChanged delivered on UI thread
- ThreadSafeObservableCollection - verify both CollectionChanged and PropertyChanged are marshalled identically

pcruzparri replied:
Same as the deadlock comment — the ThreadSafeObservableCollection has been removed from this PR, so this one no longer applies.

trishorts reviewed (commented):
Automated code review - 4 finding(s) at severity >= Major, posted as inline comments.

pcruzparri committed:
Revert out-of-scope ThreadSafeObservableCollection changes

pcruzparri committed:
Revert temporary variant description patch

pcruzparri committed:
Add guard test for protein group column count consistency

pcruzparri committed:
Use single cast PSM list for protein best-peptide metrics

pcruzparri committed:
Merge branch 'master' into ptm_stoich

trishorts committed:
Merge branch 'master' into ptm_stoich

nbollis commented on MetaMorpheus/EngineLayer/ProteinParsimony/ProteinGroup.cs:
```diff
@@ -1,145 +1,146 @@
+        /// <summary>Maps to <see cref="BioPolymerGroup.BioPolymerGroupScore"/>.</summary>
+        public double ProteinGroupScore
+        {
+            get => BioPolymerGroupScore;
```
[review-backend, review-generic, review-oop] Alias collection getters silently changed from mutable views to detached copies

`FilesForQuantification` and `IntensitiesByFile` now materialize a fresh collection on every getter call. That keeps the new base-property adapters working internally, but it also changes the public `ProteinGroup` API from reference semantics to copy semantics: any existing caller that mutates the getter result now silently mutates a throwaway copy. The PR already had to rewrite internal callers around this, so the break is real; it is just quiet for any remaining external consumers.

Suggestion: Either preserve the old mutable semantics for these alias properties, or make the break explicit by returning read-only views / renaming the APIs so callers cannot mistake them for live collections.

pcruzparri replied:
Good call. Switched both to IReadOnly (list/dictionary) and noted in the doc that the getter hands back a snapshot, so it's clear you can't mutate it in place. Nothing internal was relying on mutating the result anymore anyway.

nbollis commented on MetaMorpheus/EngineLayer/ProteinParsimony/ProteinGroup.cs:
```diff
@@ -199,36 +208,21 @@ public string GetTabSeparatedHeader()
-                bool unfractionated = FilesForQuantification.Select(p => p.Fraction).Distinct().Count() == 1;
-                bool conditionsUndefined = FilesForQuantification.All(p => string.IsNullOrEmpty(p.Condition));
-
-                // this is a hacky way to test for SILAC-labeled data...
```
[review-oop] ProteinGroup coverage no longer satisfies the inherited BioPolymerGroup coverage contract

`ProteinGroup.CalculateSequenceCoverage()` is now a hidden `new` method that fills only the legacy flat coverage lists. It never updates mzLib's cached `CoverageResult`, even though `ProteinGroup` now inherits `BioPolymerGroup` and therefore exposes the inherited `CoverageResult` property. That property lazily falls back to the base `CalculateSequenceCoverage()` implementation, but `SpectralMatch` does not implement `IHasSequenceCoverageFromFragments`, so callers using the inherited contract can observe different or emptier coverage than the `ProteinGroup` TSV output.

Suggestion: Populate `BioPolymerGroup.CoverageResult` from the MetaMorpheus-specific calculation, or move `SpectralMatch` onto `IHasSequenceCoverageFromFragments` and let the base coverage pipeline own the calculation end to end.

pcruzparri replied:
Agreed. Nothing actually reads the inherited CoverageResult yet, so I'd rather not balloon this stoich-focused PR. A follow up would further the migration from the ProteinGroup class to the BioPolymerGroup. That being said, to just address the Coverage result concern, it can be addressed by either filling CoverageResult from our calc or having SpectralMatch implement IHasSequenceCoverageFromFragments. Recording here as a future note.

nbollis reviewed (commented):
## Review Summary

Approved findings from the dashboard were revalidated against the current PR head after a freshness sync. The head commit changed since initial analysis, but both approved findings still resolved to valid inline anchors on the updated patch.

### Major (2)
- `MetaMorpheus/EngineLayer/ProteinParsimony/ProteinGroup.cs:99` Alias collection getters now return detached copies instead of live mutable collections, creating a silent behavioral break for callers that mutate getter results.
- `MetaMorpheus/EngineLayer/ProteinParsimony/ProteinGroup.cs:415` `ProteinGroup`'s custom coverage calculation no longer keeps the inherited `BioPolymerGroup.CoverageResult` contract in sync, so inherited callers can observe coverage that diverges from MetaMorpheus output.

Grouping: 2 approved findings -> 2 inline comments | 0 summary-only.

trishorts commented on MetaMorpheus/EngineLayer/ProteinParsimony/ProteinGroup.cs:
```diff
@@ -182,7 +183,15 @@ public void GetIdentifiedPeptidesOutput(List<SilacLabel> labels)
             }
         }
```
**Major** - Per-group HasIntensityData gating can desync TSV header and row columns

**Problem:**
Both `GetTabSeparatedHeader()` and `ToString()` emit the `Intensity_`/`IntensityOccupancy_` columns conditionally on `group.HasIntensityData`, evaluated per `SampleGroupResult` of the *individual* protein group. The header is produced once from a single representative `ProteinGroup`, while each data row is produced by a different `ProteinGroup` instance via its own `SampleGroupResults`. If `HasIntensityData` (or the count/identity of `SampleGroupResults`) for a given sample-group label differs between the header-source group and any row group, the row will emit a different number of tab-delimited fields than the header, misaligning the entire TSV from that column onward. The same hazard applies to the outer `if (SampleGroupResults != null)` guard: a row whose `SampleGroupResults` is null while the header group's was non-null (or vice-versa) silently drops/adds all quantification columns for that row.

**Impact:**
Column-misaligned protein TSV output: intensity, occupancy, PSM count, decoy flag, and QValue values shift into the wrong columns for affected rows, corrupting downstream parsing and any occupancy analysis. The failure is data-dependent (a protein with no intensity in one sample group), so it can pass CI yet break on real datasets.

**Suggested fix:**
Make the column shape a single property of the file/run rather than per-protein-group: decide the set of emitted columns (and each label's HasIntensityData) once from the union of all groups, and have every row honor that same shape (emitting empty fields where a particular group lacks intensity). Ensure the header is derived from the same authoritative shape used by the rows.

**Affected test scenarios:**
- WriteProteinGroupsTsv - one protein group has intensity in sample group A, another has none - all rows have identical column count equal to header
- WriteProteinGroupsTsv - header-source group's SampleGroupResults null but a later group non-null - column counts stay consistent

pcruzparri replied:
Same thing as the earlier finding. Added a test that checks every row's column count matches the header across multiple groups. The actual header refactor should go in a separate PR. It is currently fragile but fully working.

trishorts reviewed (commented):
Automated code review - 1 finding(s) at severity >= Major, posted as inline comments.

trishorts commented on MetaMorpheus/TaskLayer/GlycoSearchTask/PostGlycoSearchAnalysisTask.cs:
```diff
@@ -538,25 +545,30 @@ private void QuantificationAnalysis()
+                    // Populate SampleGroupResults from the shared spectraFileInfo so
+                    // every PG carries the same dynamic-column schema. Without this, the writer
+                    // would have no way to produce uniform headers/rows.
+                    proteinGroup.PopulateSampleGroupResults();
```
**Major** — PopulateSampleGroupResults runs even when FlashLfqResults is null, on unset intensities

`proteinGroup.PopulateSampleGroupResults()` is now called unconditionally inside `if (ProteinGroups != null)`, *outside* the `if (Parameters.FlashLfqResults != null)` guard that assigns `IntensitiesByFile`. When `DoQuantification` is true but `flashLFQIdentifications.Any()` is false (no IDs below 1% FDR), `Parameters.FlashLfqResults` stays null, so the intensity dictionary is never built — yet `PopulateSampleGroupResults()` still executes and must read `IntensitiesByFile` to produce the `Intensity_<sample>` / `IntensityOccupancy_<sample>` columns. Since this PR changed the `IntensitiesByFile` getter to return `backingField.ToDictionary(...)`, a never-assigned backing field yields either a `NullReferenceException` or all-zero intensity columns. The base code skipped this path entirely (`ProteinGroups != null && Parameters.FlashLfqResults != null`).

**Impact:** A glyco run with quantification enabled but zero quantifiable identifications can throw during protein-group result writing, or emit misleading all-zero intensity/occupancy columns that did not previously exist.

**Suggested fix:** Only call `PopulateSampleGroupResults()` after intensity state is known valid, or initialize `IntensitiesByFile` to an empty dictionary before the call when `FlashLfqResults` is null. Mirror whatever guard the `PostSearchAnalysisTask` path uses so the two quantification implementations behave identically.

pcruzparri replied:
The NRE can't really happen here — the getter is null-safe and PopulateSampleGroupResults is fine with a null intensity dict (just gives count-only columns). It was inconsistent with the search path though, so I made glyco always set IntensitiesByFile (zeros when there's no LFQ) to match. Tests still pass.

pcruzparri committed:
Return read-only snapshots from protein group quant getters

pcruzparri committed:
Always assign glyco protein group intensities for consistent columns

pcruzparri committed:
Merge branch 'ptm_stoich' of https://github.com/pcruzparri/metamorpheus into ptm_stoich

pcruzparri committed:
Populate quant columns for per-file protein group output

pcruzparri committed:
Merge branch 'master' into ptm_stoich

trishorts reviewed (dismissed):
Approving. I re-reviewed against the current head (`ae97ad2`) — not just the reply threads — and every Critical/Major finding from this round is resolved or acknowledged-and-deferred with a test:

- **Variant NRE** — reverted to master's `VariantCallFormatDataString.Description`; the incomplete guard is gone. ✅
- **`.OfType<SpectralMatch>()` dropping PSMs** — now a single `Cast<SpectralMatch>().ToList()` reused for Score/QValue/PEP, so all three are computed over the same population and an unexpected type throws loudly. ✅
- **ThreadSafeObservableCollection deadlock (Critical) + PropertyChanged marshalling** — whole (out-of-scope) class reverted out of the PR. ✅
- **Alias getters silently detached** — now `IReadOnly` with the snapshot semantics documented. ✅
- **Glyco `PopulateSampleGroupResults` on null LFQ** — glyco now always assigns `IntensitiesByFile` (zeros when no LFQ) and populates uniformly, matching the search path. ✅

Nice, responsive turnaround — the two genuinely risky/out-of-scope items (the deadlock and the variant patch) were cleanly reverted rather than papered over.

## Follow-up notes (non-blocking — for the deferred BioPolymerGroup migration PR)

Two of the findings were reasonably deferred with tests. Recording the residual debt so it isn't lost in the reply threads:

1. **Method-hiding is a trap for the very migration this defers to.** `GetTabSeparatedHeader()` and `CalculateSequenceCoverage()` are now `new`-hidden on a type that *is-a* `BioPolymerGroup`. It's safe **today** only because every caller uses the concrete type — I confirmed `WriteProteinGroupsToTsv` (`PostSearchAnalysisTask.cs:1554`) holds `List<ProteinGroup>` and calls `.First().GetTabSeparatedHeader()`, so the MM override wins. But the follow-up that moves logic onto `BioPolymerGroup` must resolve this **together with** the `CoverageResult` contract finding (#`3443992211`) — the first base-typed caller would otherwise silently get empty columns / diverging coverage.

2. **Header/row column uniformity is enforced by producer discipline, not structure.** The header is built from one representative group while each row re-derives its own columns from its own `SampleGroupResults`/`HasIntensityData`. It stays aligned only because the search and glyco paths now both assign `IntensitiesByFile` for every group (zeros for unmeasured), and `TestMultipleProteinGroupsHeaderAndRowsHaveSameColumnCount` locks that case. It remains latent TSV-corruption risk if any *future* path populates `SampleGroupResults` non-uniformly. The single-shared-schema fix (drive header + rows from one authoritative column shape) is the real resolution and should be the deferred PR — please track it as an issue, not just a reply.

3. **Trivial perf, optional:** the `IntensitiesByFile` getter allocates a fresh dictionary on every call, and around `ProteinGroup.cs:650` it's called twice to read a single value (null-check then `GetValueOrDefault`). Cache in a local. Not worth holding anything for.

None of these block the merge.

trishorts committed:
Merge branch 'master' into ptm_stoich

pcruzparri committed:
Merge branch 'master' into ptm_stoich

pcruzparri committed:
Merge branch 'master' into ptm_stoich

nbollis commented on MetaMorpheus/EngineLayer/SpectralMatch.cs:
```diff
@@ -80,26 +60,82 @@ protected SpectralMatch(IBioPolymerWithSetMods peptide, int notch, double score,
+        public IEnumerable<IBioPolymerWithSetMods> GetIdentifiedBioPolymersWithSetMods() =>
+            BestMatchingBioPolymersWithSetMods.Select(h => h.SpecificBioPolymer);
+
+        public int CompareTo(ISpectralMatch? other)
```
Careful. We already have a carefully orchestrated compare method for spectral matches, we should unify them

pcruzparri replied:
This one was a bit of a rabbithole but unified it. CompareTo(ISpectralMatch) now does score-with-tolerance, then scan number reversed, instead of the raw score comparison. Note that in instances where we are comparing by SpectralMatch type, this interface method is not called, and is just implemented for ISpectralMatch contract. So, comparing SpectralMatch objects will go through comparing score->runner up score->precursor mass error ppm->scan number. Calling a comparison on a generic ISpectralMatch (currently no references in MM) will go through this interface method and only compare by score->Scan number, since the other two fields aren't in the interface.

nbollis commented on MetaMorpheus/TaskLayer/GlycoSearchTask/PostGlycoSearchAnalysisTask.cs:
```diff
@@ -451,7 +456,8 @@ private void QuantificationAnalysis()
                         string.Join("|", proteinsOrderedByAccession.Select(p => p.Organism).Distinct()));
 
-                    foreach (var psm in proteinGroup.AllPsmsBelowOnePercentFDR.Where(v => v.FullSequence != null))
+                    foreach (var psm in proteinGroup.AllPsmsBelowOnePercentFDR.OfType<SpectralMatch>()
```
There may be issues here with the spectral matches being of the type GlycoSpectralMatch. A cast or something may be more appropriate here, unsure, but please verify this.

pcruzparri replied:
Good catch, forgot to address this on the glyco path. Fixed now.

pcruzparri committed:
Merge branch 'master' into ptm_stoich

pcruzparri committed:
Tighten ISpectralMatch handling in protein group paths
Cast instead of OfType at the four downcast sites so an unexpected
implementation throws rather than dropping out of sequence coverage.
Align CompareTo(ISpectralMatch) with the projection BaseSpectralMatch
already derives from CompareTo(SpectralMatch).

pcruzparri committed:
Merge branch 'ptm_stoich' of https://github.com/pcruzparri/metamorpheus into ptm_stoich

pcruzparri committed:
Merge branch 'master' into ptm_stoich

pcruzparri committed:
Merge branch 'master' into ptm_stoich

pcruzparri committed:
Merge branch 'master' into ptm_stoich

trishorts reviewed (commented):
Re-reviewed against the current head. My 2026-07-09 approval was dismissed by later pushes, so this supersedes it.

Context that shaped the review: the occupancy backend is already merged and shipped upstream — mzLib **#955**, itself titled "Ptm stoichiometry", landed `PositionFrequencyAnalysis`, `BioPolymerGroup.PopulateSampleGroupResults()`, `SampleGroupResult`, `ModificationOccupancyCalculator` and `ScanMetadata` in one commit, first tagged **1.0.577**. MM master pins **1.0.585**, so it is on the classpath today and no bump is needed. This PR is the consuming half of that pair, and it consumes it correctly: no local `ScanMetadata` is redefined, and `IntensitiesByFile` is a genuine adapter over the base's `IntensitiesBySample`. The direction is right. The findings below are all about fidelity at that seam, not about duplication.

Nine inline comments: six I'd consider blocking, three worth a decision.

**One thing in the description doesn't match the diff.** Item 5 describes `ThreadSafeObservableCollection` improvements "with comprehensive test coverage" — that identifier appears **zero** times in the diff, and exists in neither MetaMorpheus nor mzLib. What the diff actually does to concurrency is weaken an existing race regression test (see the comment on `FragmentReanalysisRaceConditionTest.cs`). Worth correcting the description either way.

Also of note: MM **#2187** edits the same `ProteinParsimony/ProteinGroup.cs` and will conflict with this refactor.

trishorts commented on MetaMorpheus/TaskLayer/SearchTask/PostSearchAnalysisTask.cs:
```diff
@@ -616,6 +621,16 @@ private void QuantificationAnalysis()
+                {
+                    foreach (var proteinGroup in ProteinGroups)
+                    {
+                        proteinGroup.PopulateSampleGroupResults();
```
**Major** — multiplex and non-LFQ searches now emit no PTM occupancy at all

This is the only site (with `:1053` and `PostGlycoSearchAnalysisTask.cs:571`) that calls `PopulateSampleGroupResults()`. `ProteinGroup.cs` never calls it itself — the `new`-hidden `GetTabSeparatedHeader()` bypasses the base class's lazy auto-population, and its own comment says so: *"Dynamic columns appear only when SampleGroupResults has been explicitly populated"* (`ProteinGroup.cs:221`).

But this call sits at the **end** of `QuantificationAnalysis()`, after two early returns:

```csharp
if (Parameters.SearchParameters.DoMultiplexQuantification) { ...; return; }   // :256-263
if (!Parameters.SearchParameters.DoLabelFreeQuantification) { return; }        // :266-269
```

So for TMT/iTRAQ runs and any search with LFQ off, `SampleGroupResults` stays null, and both `GetTabSeparatedHeader()` (`ProteinGroup.cs:224`) and `ToString()` (`:344`) skip every occupancy column.

Meanwhile this PR deletes the `"Modification Info List"` column and the occupancy block at the tail of `CalculateSequenceCoverage()` — which master wrote **unconditionally**, independent of quantification. Net effect: those runs go from always having per-residue occupancy to having none.

To be fair to the stated scope: the PR promises count-based columns survive when the *experimental design* is unavailable, and that path does reach this line. It's the other two early returns that aren't covered. Options are to populate count-based results before the early returns, or to keep writing the old `ModsInfo` column when `SampleGroupResults` is null.

Note the whole method is wrapped in `try/catch → EngineCrashed`, so any exception mid-quantification also skips this call and silently drops the columns.

trishorts commented on MetaMorpheus/EngineLayer/SpectralMatch.cs:
```diff
@@ -116,9 +108,55 @@ protected SpectralMatch(IBioPolymerWithSetMods peptide, int notch, double score,
+            return other.OneBasedScanNumber.CompareTo(this.ScanNumber); //reverse the comparision so that the lower scan number comes first.
+        }
+
+        public bool Equals(ISpectralMatch? other)
```
**Major** — value equality declared with no matching `GetHashCode`

This adds a value-based definition (`FullFilePath`, `ScanNumber`, `FullSequence`), but `grep -c GetHashCode` over this file returns **0** — `SpectralMatch` still inherits `object.GetHashCode()` (reference identity) and does not override `Equals(object)`.

That matters because PSMs are stored in a hash-based set keyed by the interface:

```csharp
// ProteinGroup.cs:40
AllPsmsBelowOnePercentFDR = new HashSet<ISpectralMatch>();
```

`ISpectralMatch : IEquatable<ISpectralMatch>`, so `EqualityComparer<ISpectralMatch>.Default` resolves to the generic comparer and *does* call this new `Equals` — while hashing by reference. Two PSMs equal by the definition just declared land in different buckets, so `Contains`/`UnionWith` never honour it. The equality contract is broken in the way that's hardest to see: the code compiles, the tests pass, and the set silently stops deduping by the rule it now advertises.

mzLib's sibling pairs the same definition with `GetHashCode() => HashCode.Combine(FullFilePath, OneBasedScanNumber, FullSequence)` **and** an `Equals(object)` override — worth mirroring.

There's also an asymmetry: `BaseSpectralMatch.Equals(ISpectralMatch other)` is `if (other is BaseSpectralMatch sm) return Equals(sm); return false;`. `SpectralMatch` is not a `BaseSpectralMatch`, so `mmPsm.Equals(baseSm)` can be true while `baseSm.Equals(mmPsm)` is always false.

trishorts commented on MetaMorpheus/EngineLayer/SpectralMatch.cs:
```diff
@@ -116,9 +108,55 @@ protected SpectralMatch(IBioPolymerWithSetMods peptide, int notch, double score,
+        /// Returns ReporterIonIntensities if available (isobaric), a singleton array of
+        /// PrecursorScanIntensity for LFQ, or null if neither is populated.
+        /// </summary>
+        double[]? ISpectralMatch.Intensities =>
```
**Major** — `IntensityOccupancy_*` is computed from precursor intensity, not FlashLFQ

```csharp
double[]? ISpectralMatch.Intensities =>
    IsobaricMassTagReporterIonIntensities ??
    (PrecursorScanIntensity > 0 ? new[] { PrecursorScanIntensity } : null);
```

mzLib's consumer treats a single-element array as *the* quantification intensity for that PSM:

```csharp
// ModificationOccupancyCalculator.cs:83-84 (also :125, :170, :202)
if (psm.Intensities is { Length: 1 })
    totals.totalIntensity += psm.Intensities[0];
```

and that value is what feeds `IntensityBasedStoichiometry` → the `IntensityOccupancy_<sample>` column. So the "intensity-based stoichiometry from FlashLFQ peptide intensities" the description advertises is actually derived from each PSM's raw MS1 precursor intensity — a **different data source** from the adjacent `Intensity_<sample>` column, which really does come from `IntensitiesBySample`/FlashLFQ. Two neighbouring columns in the same TSV, drawn from two different measurements.

Worse, `Ms2ScanWithSpecificMass.cs:21` does `PrecursorIntensity = precursorIntensity ?? 1`. When the deconvoluter supplied no precursor intensity, the `> 0` guard still passes and **every PSM contributes exactly 1.0** — so `IntensityOccupancy` silently degenerates into a copy of `CountOccupancy` while still being labelled as intensity-based. There's no way for a user to tell the two cases apart from the output.

If the intent is FlashLFQ intensities, this needs to read from the quantification results rather than the scan. If the precursor-intensity fallback is deliberate, it should at least distinguish "no intensity recorded" from a real 1.0.

trishorts commented on MetaMorpheus/TaskLayer/SearchTask/PostSearchAnalysisTask.cs:
```diff
@@ -551,18 +550,19 @@ private void QuantificationAnalysis()
                 foreach (var spectraFile in psmsGroupedByFile)
                 {
-                    var rawfileinfo = spectraFileInfo.First(p => p.FullFilePathWithExtension.Equals(spectraFile.Key));
+                    var rawfileinfo = spectraFileInfo.FirstOrDefault(p => p.FullFilePathWithExtension.Equals(spectraFile.Key));
```
**Major** — a file/path mismatch now silently drops every PSM from that file

`.First(...)` became `.FirstOrDefault(...)` plus a bare `continue`. Previously the mismatch threw, the outer `catch` called `EngineCrashed("Quantification", e)`, and the user got a visible warning plus a `Quantification_crash.txt`. Now every PSM from the unmatched file is dropped from `flashLFQIdentifications` with no `Warn(...)`, no status message, and no crash file.

A partial mismatch is the dangerous case: protein and peptide intensities get computed from a subset of the data and the run looks completely normal. A total mismatch leaves `Parameters.FlashLfqResults` null while the output still advertises `Intensity_*` and `IntensityOccupancy_*` columns — quantified-looking output with every intensity zero.

If skipping is genuinely the right behaviour here, it needs a `Warn($"...")` naming the file so it appears in the run log; if it isn't, the `.First()` should stay.

trishorts commented on MetaMorpheus/EngineLayer/SpectralMatch.cs:
```diff
@@ -83,28 +63,40 @@ protected SpectralMatch(IBioPolymerWithSetMods peptide, int notch, double score,
+        public int ScanNumber => ScanMetadata.OneBasedScanNumber;
+        public int? PrecursorScanNumber => ScanMetadata.OneBasedPrecursorScanNumber;
+        public double ScanRetentionTime => ScanMetadata.RetentionTime;
+        public int ScanExperimentalPeaks => ScanMetadata.NumPeaks;
```
**Major** — "Num Experimental Peaks" changes value for XCorr-preprocessed data

This now reads a snapshot taken in the `Ms2ScanWithSpecificMass` constructor. Master read it live at PSM-construction time (`SpectralMatch.cs:30`, `ScanExperimentalPeaks = scan.NumPeaks`).

The ordering is what breaks it. `Ms2ScanWithSpecificMass` objects are built inside `_GetMs2Scans` (`MetaMorpheusTask.cs:350`, `:367`, `:403`), but `XCorrPrePreprocessing` mutates the spectrum **in place, afterwards**:

```csharp
// MetaMorpheusTask.cs:466 (and :481 for child scans) — after _GetMs2Scans returns at :434
parentScan.TheScan.MassSpectrum.XCorrPrePreprocessing(0, 1969, parentScan.TheScan.IsolationMz.Value);
```

So the snapshot records the **pre**-processing peak count where the old code recorded the **post**-processing count. For the affected dissociation path the column in `AllPSMs.psmtsv` silently changes value against previous releases, and no test asserts it.

Capturing `NumPeaks` lazily, or re-snapshotting after preprocessing, would preserve the old value.

trishorts commented on MetaMorpheus/Test/ProteinGroupTest.cs:
```diff
@@ -244,12 +413,82 @@ public static void TestModificationInfoListInProteinGroupsOutput()
+            pg.GetIdentifiedPeptidesOutput(null);
+            var tsv1 = pg.ToString();
+            Assert.That(tsv1, Does.Contain(pwsm1.BaseSequence));
+            Assert.That(tsv1.Split('\t')[4], Does.Not.Contain("[")); // unique-peptides column has no mod notation
```
**Major (test)** — this asserts on the mass column, not the unique-peptides column

`GetTabSeparatedHeader()` orders columns as Accession(0), Gene(1), Organism(2), Protein Full Name(3), **Protein Unmodified Mass(4)**, Number of Proteins in Group(5), **Unique Peptides(6)**.

Index `[4]` is the unmodified-mass list, so `Does.Not.Contain("[")` is being applied to a `"|"`-joined list of doubles and is true no matter what `GetIdentifiedPeptidesOutput` produced. The two `Is.Not.Empty` checks in branches 3 and 4 have the same problem.

Three of the four branch assertions in this new test can therefore never fail — the SILAC label branches end up covered only by `Assert.DoesNotThrow`. Index `[6]` is the column the comment describes.

Given the column set is now dynamic, indexing by name (find the header offset for `"Unique Peptides"`) would be more durable than a literal index.

trishorts commented on MetaMorpheus/Test/MetaDraw/FragmentReanalysisRaceConditionTest.cs:
```diff
@@ -47,39 +45,53 @@ public static void MatchIonsWithNewTypes_ConcurrentCalls_DoNotThrow()
-                Directory.Delete(outputFolder, true);
-            }
+            // Wait for matcher to finish (or cancel)
+            matcherTask.Wait(TimeSpan.FromSeconds(10));
```
**Please reconsider** — this weakens a race regression test with no accompanying production change

The rewrite drops `[NonParallelizable]`, collapses `Enumerable.Range(0, Environment.ProcessorCount)` concurrent callers down to a single matcher plus one mutator, narrows `catch (Exception)` to `catch (InvalidOperationException)`, and removes the `try/finally` that cleaned `outputFolder`.

The deciding line is this one:

```csharp
matcherTask.Wait(TimeSpan.FromSeconds(10));   // return value discarded
```

The previous version asserted on it — `Assert.That(completed, Is.True, "Concurrent rematching did not complete within the timeout.")`. `Task.Wait(TimeSpan)` returns `false` on timeout rather than throwing, so a hang or deadlock now passes silently. Combined with dropping the `ProcessorCount` fan-out, the test is unlikely to reproduce the race it was written for.

`FragmentationReanalysisViewModel.cs` isn't in this PR, so nothing in the production code motivates the change. If the old test was flaky, that's worth fixing on its own terms — but as it stands this removes coverage without replacing it. At minimum, restore the `Is.True` assertion on the `Wait` result.

trishorts commented on MetaMorpheus/EngineLayer/SpectralMatch.cs:
```diff
@@ -462,22 +506,12 @@ protected SpectralMatch(SpectralMatch psm, List<SpectralMatchHypothesis> bestMat
-            FullFilePath = psm.FullFilePath;
-            ScanIndex = psm.ScanIndex;
             FdrInfo = psm.FdrInfo;
+            PeptideFdrInfo = psm.PeptideFdrInfo;
```
**Question** — this line is new, and unrelated to the stated purpose

The copy-constructor previously copied `FdrInfo` but not `PeptideFdrInfo`. Adding it changes peptide-level FDR propagation through every `CrosslinkSpectralMatch` / `GlycoSpectralMatch` clone, and through the SILAC clone path.

That may well be a genuine bug fix — but it isn't mentioned in the description, no test covers it, and FDR propagation is not somewhere a silent change is comfortable. Could you confirm it's deliberate, and say what it fixes? If it is a fix, it deserves its own test; if it's incidental, it should probably come out of this PR.

trishorts commented on MetaMorpheus/EngineLayer/ProteinParsimony/ProteinGroup.cs:
```diff
@@ -736,23 +641,32 @@ public ProteinGroup ConstructSubsetProteinGroup(string fullFilePath, List<SilacL
             }
 
-            if (IntensitiesByFile == null)
+            if (IntensitiesByFile == null || spectraFileInfo == null)
```
**Performance** — the copying getters are re-evaluated several times per call here

`FilesForQuantification` and `IntensitiesByFile` are no longer fields; each read allocates a fresh collection:

```csharp
get => SamplesForQuantification?.OfType<SpectraFileInfo>().ToList();          // :105
get => IntensitiesBySample?.ToDictionary(kvp => (SpectraFileInfo)kvp.Key, ...); // :116
```

Within `ConstructSubsetProteinGroup`, `FilesForQuantification` is read four times (the null check, the main lookup, and both SILAC fake-path lookups) and `IntensitiesByFile` twice — this null check and the `GetValueOrDefault` just below. Each is a full O(files) copy.

The method is called once per protein group per spectra file (`PostSearchAnalysisTask.cs:1038`), so total copying is on the order of `files² × proteinGroups` — a path that previously allocated nothing, because these were plain fields.

Hoisting each getter into a local at the top of the method fixes the hot path cheaply. Worth noting more broadly: eager copy-on-read getters are new to this codebase — the nearest precedent (`SpectralMatch.cs:230`) is lazy and still carries the author's own comment worrying about its per-call cost.

trishorts replied:
**Performance** — the copying getters are re-evaluated several times per call here

`FilesForQuantification` and `IntensitiesByFile` are no longer fields; each read allocates a fresh collection:

```csharp
get => SamplesForQuantification?.OfType<SpectraFileInfo>().ToList();          // :105
get => IntensitiesBySample?.ToDictionary(kvp => (SpectraFileInfo)kvp.Key, ...); // :116
```

Within `ConstructSubsetProteinGroup`, `FilesForQuantification` is read four times (the null check, the main lookup, and both SILAC fake-path lookups) and `IntensitiesByFile` twice — this null check and the `GetValueOrDefault` just below. Each is a full O(files) copy.

The method is called once per protein group per spectra file (`PostSearchAnalysisTask.cs:1038`), so total copying is on the order of `files² × proteinGroups` — a path that previously allocated nothing, because these were plain fields.

Hoisting each getter into a local at the top of the method fixes the hot path cheaply. Worth noting more broadly: eager copy-on-read getters are new to this codebase — the nearest precedent (`SpectralMatch.cs:230`) is lazy and still carries the author's own comment worrying about its per-call cost.

pcruzparri committed:
Use reference equality for ISpectralMatch to match the inherited hash code

pcruzparri committed:
Snapshot the quantification adapters once per subset protein group

pcruzparri committed:
Warn when a spectra file has no quantification entry instead of dropping it silently

pcruzparri committed:
Report count-based occupancy when quantification is off or unavailable

pcruzparri committed:
Weight intensity occupancy by FlashLFQ peptide areas instead of precursor intensity

pcruzparri committed:
Re-read the scan peak count after XCorr rewrites the spectrum

pcruzparri committed:
Resolve the unique-peptides column by header name instead of a fixed index

pcruzparri committed:
Restore the concurrent-rematching test from master

pcruzparri committed:
Merge branch 'master' into ptm_stoich

nbollis commented on MetaMorpheus/TaskLayer/SearchTask/PostSearchAnalysisTask.cs:
```diff
@@ -254,6 +253,12 @@ private void QuantificationAnalysis()
+                // before every return below and before anything that can throw. Without an experimental
+                // design the columns are labelled per spectra file; the label-free path repopulates
+                // them with intensities once FlashLFQ has run.
+                PopulateCountBasedOccupancy();
```
Should count based occupancy only run if we are doing quantification? Or should count based occupancy be run regardless of if we are doing quantification?
